x/superfluid/keeper/stake.go

```diff
@@ -656,3 +658,150 @@ func (k Keeper) IterateDelegations(ctx sdk.Context, delegator sdk.AccAddress, fn
 		fn(index+int64(i), delegation)
 	}
 }
+
+// UnbondConvertAndStake converts given lock to osmo and stakes it to given validator.
+// Supports conversion of 1)superfluid bonded 2)superfluid undelegating 3)vanilla unlocking 4) unlocked locks.
+// If valAddr is empty, we attempt to get staking preference from valset-pref module and stake to the given validator.
+func (k Keeper) UnbondConvertAndStake(ctx sdk.Context, lockID uint64, sender, valAddr string,
+	minAmtToStake sdk.Int) (totalAmtConverted sdk.Int, totalSharesDelegated sdk.Dec, err error) {
+	senderAddr, err := sdk.AccAddressFromBech32(sender)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	// use routeMigration method to check status of lock(either superfluid staked, superfluid unbonding, vanialla locked, unlocked)
+	_, migrationType, err := k.routeMigration(ctx, senderAddr, int64(lockID))
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	// if superfluid bonded, first change it into superfluid undelegate to burn minted osmo and instantly undelegate.
+	if migrationType == SuperfluidBonded {
+		_, err = k.undelegateCommon(ctx, sender, lockID)
+		if err != nil {
+			return sdk.ZeroInt(), sdk.ZeroDec(), err
+		}
+	}
+
+	if migrationType == SuperfluidBonded || migrationType == SuperfluidUnbonding || migrationType == NonSuperfluid {
+		totalAmtConverted, totalSharesDelegated, err = k.convertLockToStake(ctx, senderAddr, valAddr, lockID, minAmtToStake)
+		if err != nil {
+			return sdk.ZeroInt(), sdk.ZeroDec(), err
+		}
+	} else { // liquid gamm shares without locks are not supported
+		return sdk.ZeroInt(), sdk.ZeroDec(), fmt.Errorf("unsupported staking conversion type")
+	}
+
+	return totalAmtConverted, totalSharesDelegated, nil
+}
+
+// convertLockToStake handles locks that are superfluid bonded, superfluid unbonding, vanilla locked(unlocking) locks.
+// Deletes all associated state, converts the lock itself to staking delegation by going through exit pool and swap.
+func (k Keeper) convertLockToStake(ctx sdk.Context, sender sdk.AccAddress, valAddr string, lockId uint64,
+	minAmtToStake sdk.Int) (totalAmtConverted sdk.Int, shares sdk.Dec, err error) {
+	lock, err := k.lk.GetLockByID(ctx, lockId)
+	if err != nil {
+		return sdk.ZeroInt(), sdk.ZeroDec(), err
+	}
+
+	lockCoin := lock.Coins[0]
+	poolIdLeaving, err := gammtypes.GetPoolIdFromShareDenom(lockCoin.Denom)
```

Comment:
Can we always get the pool id by using this function? What if this is a CL lock denom?

I'm referring to CL lock denom as made here:
https://github.com/osmosis-labs/osmosis/blob/1bef46333cb2d4b3464f6779e33bea9275e1e41d/x/concentrated-liquidity/position.go#L485

Reply:
AFAIK, we are not supporting CL yet. In that sense, added double check + test case to ensure lock has denom of balancer pool in https://github.com/osmosis-labs/osmosis/pull/5949/commits/be4c03a5c093354abdd82edf83f16dc2d38d5bfd

Reply:
Do we have a test case that attempts to migrate a lock that has a non gamm token prefix?

Reply:
Yes, this is the case
https://github.com/osmosis-labs/osmosis/blob/b953043f91b328699ff62f2f4f4c84c3e978c9c6/x/superfluid/keeper/stake_test.go#L1152-L1157